client/src/test/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClientTest.java

```diff
@@ -15,260 +15,269 @@
+        .andReturn(new io.confluent.kafka.schemaregistry.client.rest.entities.Schema(SUBJECT_0,
                                                                                      version,
-                                                                                     id,
-                                                                                     schema));
+            ID_25,
+            SCHEMA_STR_0));
 
     expect(restService.deleteSchemaVersion(RestService.DEFAULT_REQUEST_PROPERTIES,
-                                           subject,
+        SUBJECT_0,
                                            String.valueOf(version)))
         .andReturn(0);
 
     replay(restService);
 
-    assertEquals(id, client.register(subject, avroSchema));
-    assertEquals(version, client.getVersion(subject, avroSchema));
-    assertEquals(version, client.getVersion(subject, avroSchema)); // hit the cache
+    assertEquals(ID_25, client.register(SUBJECT_0, AVRO_SCHEMA_0));
+    assertEquals(version, client.getVersion(SUBJECT_0, AVRO_SCHEMA_0));
+    assertEquals(version, client.getVersion(SUBJECT_0, AVRO_SCHEMA_0)); // hit the cache
 
-    assertEquals(Integer.valueOf(0), client.deleteSchemaVersion(subject, String.valueOf(version)));
+    assertEquals(Integer.valueOf(0),
+        client.deleteSchemaVersion(SUBJECT_0, String.valueOf(version)));
 
     verify(restService);
   }
+
+  @Test
+  public void testDeleteVersionNotInVersionCache() throws Exception {
+    expect(client.deleteSchemaVersion(Collections.emptyMap(), SUBJECT_0, "0"))
+        .andReturn(10);
+
+    replay(restService);
+
+    final Integer result = client.deleteSchemaVersion(Collections.emptyMap(), SUBJECT_0, "0");
+
+    assertEquals((Integer)10, result);
+    verify(restService);
+  }
+
+  @Test(expected = NullPointerException.class)
+  public void testDeleteNullSubjectThrows() throws Exception {
+    client.deleteSubject(null);
+  }
+
+  @Test
+  public void testThreadSafe() throws Exception {
+    expect(restService.registerSchema(anyString(), eq(SUBJECT_0)))
+        .andReturn(ID_25)
+        .anyTimes();
+
+    expect(restService.getId(ID_25))
+        .andReturn(new SchemaString(SCHEMA_STR_0))
+        .anyTimes();
+
+    expect(restService.lookUpSubjectVersion(eq(AVRO_SCHEMA_0.toString()), eq(SUBJECT_0), anyBoolean()))
+        .andReturn(SCHEMA_DETAILS)
+        .anyTimes();
+
+    replay(restService);
+
+    IntStream.range(0, 1_000)
+        .parallel()
+        .forEach(idx -> {
+          try {
+            final int id = client.register(SUBJECT_0, AVRO_SCHEMA_0);
+            final int version = client.getVersion(SUBJECT_0, AVRO_SCHEMA_0);
+            client.getId(SUBJECT_0, AVRO_SCHEMA_0);
+            client.getBySubjectAndId(SUBJECT_0, id);
+            client.deleteSchemaVersion(SUBJECT_0, String.valueOf(version));
+            client.deleteSubject(SUBJECT_0);
+          } catch (final IOException | RestClientException e) {
+            //
```

Comment:
I know these are not expected, but wouldn't it be better to wrap these in a runtime exception to propagate them out of the forEach?

Also, would it be better to check the return of these methods to make sure they're also functioning correctly?

Reply:
Yep, fair enough.